### PR TITLE
add pprof config and start up logic

### PR DESCRIPTION
**Pull request by wxing1292**

### CODE REVIEW

File: common/pprof.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+type (
+	// PProfInitializer initialize the pprof based on config
+	PProfInitializer interface {
```
Comment (madhuravi) [inline common/pprof.go]: Do we need an interface for this? We will likely only ever have one implementation. We can just use the struct directly.

File: common/service/config/pprof.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+// NewInitializer create a new instance of PProf Initializer
+func (cfg *PProf) NewInitializer(logger bark.Logger) *PProfInitializerImpl {
+	return &PProfInitializerImpl{
+		PProf:  cfg,
+		Logger: logger,
+	}
+}
+
+// Initialize the pprof based on config
+func (initializer *PProfInitializerImpl) Initialize() error {
```
Comment (madhuravi) [inline common/service/config/pprof.go]: nit: We often use "Start" for this and have that method on a lot of our structures. I would recommend naming this Start as well to keep it consistent.

File: common/service/config/pprof.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+type (
+	// PProfInitializerImpl initialize the pprof based on config
+	PProfInitializerImpl struct {
+		PProf  *PProf
+		Logger bark.Logger
+	}
+)
+
+// NewInitializer create a new instance of PProf Initializer
+func (cfg *PProf) NewInitializer(logger bark.Logger) *PProfInitializerImpl {
```
Comment (madhuravi) [inline common/service/config/pprof.go]: I would rather have a method NewInitializer(cfg *PProf, logger bark.Logger). That's how we do it everywhere and not much reason for this to be a member of PProf.

Comment (madhuravi) [reply common/service/config/pprof.go]: I see why you did this now w.r.t other members in the config. The reason this one seems weird is PProf is a config struct. We use config to create a NewInitializer, seems to fit more as a parameter using which a new initializer is created.

File: host/onebox.go [hunk 4]
```diff
@@ -146,10 +152,26 @@ func (c *cadenceImpl) HistoryServiceAddress() []string {
...
 
+func (c *cadenceImpl) HistoryPProfPort() []int {
+	ports := []int{}
+	startPort := 7300
+	for i := 0; i < c.numberOfHistoryHosts; i++ {
+		port := startPort + i
+		ports = append(ports, port)
+	}
+
+	c.logger.Infof("History pprof prots: %v", ports)
```
Comment (madhuravi) [inline host/onebox.go]: nit: typo prots->ports

Comment (madhuravi) [reply host/onebox.go]: Do we want to log this? Other ones aren't being logged.

File: common/service/config/pprof.go (hunk 2 above)
Comment (wxing1292) [reply common/service/config/pprof.go]: sure

File: host/onebox.go (hunk 4 above)
Comment (wxing1292) [reply host/onebox.go]: the server, when starts, actually shows the address and port:
INFO[2017-12-21T15:23:09-08:00] Created RPC dispatcher for 'cadence-matching' and listening at '127.0.0.1:7935' 
INFO[2017-12-21T15:23:09-08:00] Created RPC dispatcher for 'cadence-frontend' and listening at '127.0.0.1:7933' 
INFO[2017-12-21T15:23:09-08:00] Created RPC dispatcher for 'cadence-history' and listening at '127.0.0.1:7934'

File: common/service/service.go [hunk 5]
```diff
@@ -116,6 +119,10 @@ func (h *serviceImpl) Start() {
 	h.metricsScope.Counter(metrics.RestartCount).Inc(1)
 	h.runtimeMetricsReporter.Start()
 
+	if err := h.pprofInitializer.Initialize(); err != nil {
```
Comment (samarabbas) [inline common/service/service.go]: You probably want to allow service startup if the pprof initializer is not set.

File: common/service/config/pprof.go (hunk 2 above)
Comment (wxing1292) [reply common/service/config/pprof.go]: changed to start

File: host/onebox.go (hunk 4 above)
Comment (wxing1292) [reply host/onebox.go]: typo updated

File: common/service/service.go (hunk 5 above)
Comment (wxing1292) [reply common/service/service.go]: fixed

File: common/service/config/pprof.go (hunk 3 above)
Comment (wxing1292) [reply common/service/config/pprof.go]: in common/service, rpc.go and ringpop.go are using the same pattern.....
seems i should follow the same pattern....

**Pull request merged**